### PR TITLE
Fold EngineSearcher into Engine.Searcher

### PR DESCRIPTION
**Pull request by s1monw**
EngineSearcher can be easily folded into Engine.Searcher which removes
a level of inheritance that is necessary for most of it's subclasses.
This change folds it into Engine.Searcher and removes the dependency on
ReferenceManager.

### CODE REVIEW

**Review by bleskes — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 1]
```diff
@@ -1167,40 +1176,67 @@ default void onFailedEngine(String reason, @Nullable Exception e) {
...
 
-        public IndexSearcher searcher() {
+        public final IndexSearcher searcher() {
             return searcher;
         }
 
         @Override
         public void close() {
-            // Nothing to close here
+            if (!released.compareAndSet(false, true)) {
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/index/engine/Engine.java]: OMG `== false`! 😱

**Review by jasontedor — CHANGES_REQUESTED** [review]
:trollface:

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/engine/Engine.java]: You can push a fix for this oversight directly, dismiss my disapproving review, and merge. 😇

Comment (s1monw) [reply server/src/main/java/org/elasticsearch/index/engine/Engine.java]: I pushed a commit with an appropriate commit message

**Pull request merged**